microsoft/testsuites/core/msr.py

```diff
@@ -0,0 +1,91 @@
+
+# See docs for hypercall spec, sharing os info
+# is required before making hypercalls.
+#  https://learn.microsoft.com/en-us/virtualization/hyper-v-on-windows/tlfs/hypercall-interface#guest-os-identity-for-proprietary-operating-systems
+# build info: bits 0-15
+BUILD_INFO = 0x0000_0000_0000_FFFF
+# service pack, bits 16-23
+SERVICE_PACK = 0x0000_0000_00FF_0000
+# minor version, bits 24-31
+MINOR_VERSION = 0x0000_0000_FF00_0000
+# major version, bits 32-39
+MAJOR_VERSION = 0x0000_00FF_0000_0000
+# OS ID  bits 40-47
+OS_ID = 0x0000_FF00_0000_0000
+# vendor ID: bits 48-62
+VENDOR_ID = 0x7FFF_0000_0000_0000
+# the all important 'is open source os' flag.
+# indicates whether is *nix or other non-winodws
+# os.  the final bit, bit 63
+IS_OPEN_SOURCE_OS = 0x8000_0000_0000_0000
+
+
+@TestSuiteMetadata(
+    area="core",
+    category="functional",
+    description="""
+    Test suite verifies hyper-v platform id is set correctly via hypercall to host.
+    """,
+    requirement=simple_requirement(supported_os=[Linux]),
+)
+class Msr(TestSuite):
+    @TestCaseMetadata(
+        description="""
+            verify platform id is accurate in msr register
+        """,
+        priority=1,
+    )
+    def verify_hyperv_platform_id(self, node: RemoteNode) -> None:
+        distro = node.os
+        if isinstance(distro, Fedora):
+            distro.install_epel()
+        elif isinstance(distro, (Debian, Suse, CBLMariner)):
+            # no special setup, same package name
+            pass
+        else:
+            raise SkippedException("MSR platform id test not yet supported on this OS.")
+
+        distro.install_packages("msr-tools")
+        node.tools[Modprobe].load("msr")
+        id_information = node.execute(
+            "rdmsr 0x40000000",
+            shell=True,
+            sudo=True,
+            expected_exit_code=0,
+            expected_exit_code_failure_message=(
+                "Could not run rdmsr and fetch platform id info"
+            ),
+        ).stdout
+        node.log.info(f"MSR register contained OS information: {id_information}")
+        msr_register_content = int(id_information, 16)
+        os_vendor_id = (msr_register_content & VENDOR_ID) >> 48
+        node.log.info(f"Found Vendor ID: {os_vendor_id}")
+        build = msr_register_content & BUILD_INFO
+        major = (msr_register_content & MAJOR_VERSION) >> 32
+        minor = (msr_register_content & MINOR_VERSION) >> 24
+        patch = (msr_register_content & SERVICE_PACK) >> 16
+        os_id = (msr_register_content & OS_ID) >> 40
+        node.log.info(f"OS ID: {hex(os_id)}")
+        node.log.info(
+            "Found os build info: (major, minor, patch, build) "
+            f"{major}.{minor}.{patch}-{build}"
+        )
+        assert_that(msr_register_content & IS_OPEN_SOURCE_OS).described_as(
+            "OS_TYPE not set to OPEN_SOURCE in bitfield. "
+            f"Expected {hex(msr_register_content)} & {hex(IS_OPEN_SOURCE_OS)} != 0."
```

Comment:
can we make the description easy to understand?

Reply:
I'm not sure? It's pretty simple in C terms, just parsing a bitfield. I wrapped the properties into a class which will pretty print, so hopefully it's more clear what's happening

Reply:
There's extra code because I'm sure at some point we'll want to check the other fields. Right now we just need to check the IsOpenSource field for this test. This first OsPlatformInfo struct is used in the first hypercall, all of it is optional but if people mess the fields up they won't get nice things like VM health info.